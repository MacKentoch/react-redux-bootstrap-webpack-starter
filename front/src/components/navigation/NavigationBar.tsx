import React, { useState } from 'react';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
} from 'reactstrap';
import { RouteComponentProps } from 'react-router';
import { OwnProps, ReduxConnectedProps } from './index';

// #region types
type Props = RouteComponentProps & OwnProps & ReduxConnectedProps;
// #endregion

function NavigationBar({
  brand,
  navModel: { rightLinks },
  // leftNavItemClick,
  // rightNavItemClick,
  isAuthenticated,
  history,
  disconnectUser,
}: Props) {
  const [isOpen, setIsOpen] = useState(false);

  // #region navigation bar toggle
  const toggle = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event?.preventDefault();
    setIsOpen(!isOpen);
  };
  // #endregion

  // #region handlesNavItemClick event
  const handlesNavItemClick =
    (link = '/') =>
    (event: React.MouseEvent<HTMLAnchorElement>) => {
      event?.preventDefault();
      history.push(link);
    };
  // #endregion

  // #region disconnect
  const handlesDisconnect = (event: React.MouseEvent<HTMLAnchorElement>) => {
    event?.preventDefault();
    disconnectUser();
    history.push('/');
  };
  // #endregion

  return (
    <Navbar color="light" light expand="md">
      <NavbarBrand href="/">{brand}</NavbarBrand>
      <NavbarToggler onClick={toggle} />
      <Collapse isOpen={isOpen} navbar>
        <Nav className="ml-auto" navbar>
          {rightLinks.map(({ label, link }, index) => (
            <NavItem key={`${index}`}>
              <NavLink href="#" onClick={handlesNavItemClick(link)}>
                {label}
              </NavLink>
            </NavItem>
          ))}
          {isAuthenticated && (
            <NavItem>
              <NavLink href="#" onClick={handlesDisconnect}>
                Disconnect
              </NavLink>
            </NavItem>
          )}
        </Nav>
      </Collapse>
    </Navbar>
  );
}

NavigationBar.displayName = 'NavigationBar';

export default NavigationBar;
