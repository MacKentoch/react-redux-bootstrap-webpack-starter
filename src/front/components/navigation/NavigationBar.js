// @flow

// #region imports
import React, { PureComponent } from 'react';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
} from 'reactstrap';
import { withRouter } from 'react-router-dom';
import {
  type Match,
  type Location,
  type RouterHistory,
} from 'react-router-dom';
import { type AuthContextProps } from '../../contexts/auth/consumerHOC';
// #endregion

// #region flow types
type Props = {
  // withRouter HOC:
  match: Match,
  location: Location,
  history: RouterHistory,

  // parent props:
  brand: string,
  handleLeftNavItemClick: OnLeftNavButtonClick,
  handleRightNavItemClick: OnRightNavButtonClick,
  navModel: {
    leftLinks: Array<LeftLink>,
    rightLinks: Array<RightLink>,
  },

  ...any,
} & AuthContextProps;

type State = {
  isOpen: boolean,
  ...any,
};
// #endregion

class NavigationBar extends PureComponent<Props, State> {
  static defaultProps = {
    brand: 'brand',
  };

  state = {
    isOpen: false,
  };

  // #region lifecycle
  render() {
    const {
      brand,
      navModel: { rightLinks },
      isAuthenticated,
    } = this.props;

    const { isOpen } = this.state;

    return (
      <Navbar color="light" light expand="md">
        <NavbarBrand href="/">{brand}</NavbarBrand>
        <NavbarToggler onClick={this.toggle} />
        <Collapse isOpen={isOpen} navbar>
          <Nav className="ml-auto" navbar>
            {rightLinks.map(({ label, link, viewName }, index) => (
              <NavItem key={`${index}`}>
                <NavLink href="#" onClick={this.handlesNavItemClick(link)}>
                  {label}
                </NavLink>
              </NavItem>
            ))}
            {isAuthenticated && (
              <NavItem>
                <NavLink href="#" onClick={this.handlesDisconnect}>
                  Disconnect
                </NavLink>
              </NavItem>
            )}
          </Nav>
        </Collapse>
      </Navbar>
    );
  }
  // #endregion

  // #region navigation bar toggle
  toggle = (evt: SyntheticEvent<>) => {
    if (evt) {
      evt.preventDefault();
    }
    this.setState((isOpen: prevIsOpened) => ({ isOpen: !prevIsOpened }));
  };
  // #endregion

  // #region handlesNavItemClick event
  handlesNavItemClick = (link: string = '/') => (evt: SyntheticEvent<>) => {
    if (evt) {
      evt.preventDefault();
    }
    const { history } = this.props;
    history.push(link);
  };
  // #endregion

  // #region disconnect
  handlesDisconnect = (evt: SyntheticEvent<>) => {
    if (evt) {
      evt.preventDefault();
    }
    const { history, disconnectUser } = this.props;

    disconnectUser();
    history.push('/');
  };
  // #endregion
}

export default withRouter(NavigationBar);
