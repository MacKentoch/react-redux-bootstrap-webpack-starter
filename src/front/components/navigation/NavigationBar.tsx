import React, { useState } from 'react';
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
} from 'reactstrap';
import { RouteComponentProps } from 'react-router';
import { OwnProps, MappedDispatchToProps, MappedStateToProps } from './index';

// #region types
type Props = {} & RouteComponentProps &
  OwnProps &
  MappedStateToProps &
  MappedDispatchToProps;
// #endregion

function NavigationBar({
  brand,
  navModel: { rightLinks },
  leftNavItemClick,
  rightNavItemClick,
  isAuthenticated,
  history,
  disconnectUser,
}: Props) {
  const [isOpen, setIsOpen] = useState(false);

  // #region navigation bar toggle
  const toggle = (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    event && event.preventDefault();
    setIsOpen(!isOpen);
  };
  // #endregion

  // #region handlesNavItemClick event
  const handlesNavItemClick = (link: string = '/') => (
    event: React.MouseEvent<HTMLButtonElement>,
  ) => {
    event && event.preventDefault();
    history.push(link);
  };
  // #endregion

  // #region disconnect
  const handlesDisconnect = (event: React.MouseEvent<HTMLButtonElement>) => {
    event && event.preventDefault();
    disconnectUser();
    history.push('/');
  };
  // #endregion

  return (
    <Navbar color="light" light expand="md">
      <NavbarBrand href="/">{brand}</NavbarBrand>
      <NavbarToggler onClick={toggle} />
      <Collapse isOpen={isOpen} navbar>
        <Nav className="ml-auto" navbar>
          {rightLinks.map(({ label, link }, index) => (
            <NavItem key={`${index}`}>
              <NavLink href="#" onClick={handlesNavItemClick(link)}>
                {label}
              </NavLink>
            </NavItem>
          ))}
          {isAuthenticated && (
            <NavItem>
              <NavLink href="#" onClick={handlesDisconnect}>
                Disconnect
              </NavLink>
            </NavItem>
          )}
        </Nav>
      </Collapse>
    </Navbar>
  );
}

NavigationBar.displayName = 'NavigationBar';

export default NavigationBar;
