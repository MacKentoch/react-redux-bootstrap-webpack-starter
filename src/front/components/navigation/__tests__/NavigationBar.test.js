// @flow

// #region imports
import React from 'react';
import renderer from 'react-test-renderer'; // needed both for snpashot testing but also to prevent errors from enzyme
import { MemoryRouter } from 'react-router';
import NavigationBar from '../NavigationBar';
// #endregion

describe('NavigationBar component', () => {
  it('renders as expected', () => {
    const props = {
      brand: 'test',
      handleLeftNavItemClick: () => {},
      handleRightNavItemClick: () => {},
      navModel: {
        leftLinks: [
          {
            link: '/',
            label: 'home',
            viewName: 'home',
            onClick: () => {},
          },
        ],
        rightLinks: [
          {
            link: '/',
            label: 'home',
            viewName: 'home',
            onClick: () => {},
          },
        ],
      },
    };

    const component = renderer
      .create(
        <div>
          <MemoryRouter>
            <NavigationBar {...props} />
          </MemoryRouter>
        </div>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
