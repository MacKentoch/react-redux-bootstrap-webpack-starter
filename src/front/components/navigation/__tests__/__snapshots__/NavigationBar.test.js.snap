// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar component renders as expected 1`] = `
<div>
  <nav
    className="navbar navbar-expand-md navbar-light bg-light"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      test
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      style={Object {}}
    >
      <ul
        className="ml-auto navbar-nav"
      >
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
            home
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
